protocols/kad/src/behaviour.rs

```diff
@@ -2492,6 +2547,37 @@ pub enum KademliaEvent {
     PendingRoutablePeer { peer: PeerId, address: Multiaddr },
 }
 
+/// Information about progress events.
+#[derive(Debug, Clone)]
+pub struct ProgressStep {
+    /// The index into the event
+    pub count: NonZeroUsize,
+    /// Is this the final event?
+    pub last: bool,
+}
+
+impl ProgressStep {
+    fn first() -> Self {
+        Self {
+            count: NonZeroUsize::new(1).expect("1 to be greater than 0."),
+            last: false,
+        }
+    }
+
+    fn first_and_last() -> Self {
+        let mut first = ProgressStep::first();
+        first.last = true;
+        first
+    }
+
+    fn next(&self) -> Self {
+        assert!(!self.last);
+        let count = NonZeroUsize::new(self.count.get() + 1).expect("Not to exhaust usize space.");
```

Comment:
Nit pick: I don't think `+` in Rust overflows but instead panics, thus the `expect` is actually not checking for `exhaust`.

``` rust
#[allow(arithmetic_overflow)]

fn main() {
    assert_eq!(255_u8 + 1, 0);
}
```

```
Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 1.61s
     Running `target/debug/playground`
thread 'main' panicked at 'attempt to add with overflow', src/main.rs:4:16
note: [run with `RUST_BACKTRACE=1` environment variable to display a backtrace](https://play.rust-lang.org/#)
```

```suggestion
        let count = NonZeroUsize::new(self.count.get() + 1).expect("Adding 1 not to result in 0.");
```

Reply:
It's a bit more complicated, overflow checks are only inserted in `debug` builds, in `release` builds this will wrap around ( and the above program you posted will pass)